### PR TITLE
[GCS]Use new flag to indicate whether resources are updated and update realtime resources view

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?

this pr do 2 things:
1. realtime resources view in gcs node manager relies on heartbeat. when light heartbeat enabled we should handle the situation.
2. Use two new flags to indicate whether available resources or resource load is updated, as they might be updated to empty map, so only checking if the map in heartbeat is empty is not enough.

## Related issue number

part of https://github.com/ray-project/ray/issues/10355

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/constants.h [hunk 1]
```diff
@@ -42,3 +42,5 @@ constexpr char kWorkerRayletConfigPlaceholder[] = "RAY_WORKER_RAYLET_CONFIG_PLAC
 /// Public DNS address which is is used to connect and get local IP.
 constexpr char kPublicDNSServerIp[] = "8.8.8.8";
 constexpr int kPublicDNSServerPort = 53;
+
+constexpr char kResourcesTurningEmpty[] = "_TURNING_EMPTY_";
```
Comment (ffbin) [inline src/ray/common/constants.h:46]: add a comment

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 2]
```diff
@@ -459,9 +460,21 @@ void GcsNodeManager::StartNodeFailureDetector() {
...
+        auto resources_available = MapFromProtobuf(heartbeat.resources_available());
+        cluster_realtime_resources_[node_id] =
+            std::make_shared<ResourceSet>(resources_available);
+      }
+    }
+  } else {
+    auto resources_available = MapFromProtobuf(heartbeat.resources_available());
+    cluster_realtime_resources_[node_id] =
+        std::make_shared<ResourceSet>(resources_available);
+  }
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:477]: if (light_heartbeat_enabled_ && heartbeat.resources_available().contains(kResourcesTurningEmpty)) {
  cluster_realtime_resources_[node_id] = std::make_shared<ResourceSet>();
} else {
  auto resources_available = MapFromProtobuf(heartbeat.resources_available());
    cluster_realtime_resources_[node_id] =
        std::make_shared<ResourceSet>(resources_available);
}

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 3]
```diff
@@ -165,6 +165,7 @@ GcsNodeManager::GcsNodeManager(boost::asio::io_service &main_io_service,
             });
           })),
       node_failure_detector_service_(node_failure_detector_io_service),
+      light_heartbeat_enabled_(RayConfig::instance().light_heartbeat_enabled()),
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:168]: light_heartbeat_enabled_ is not need, we can use RayConfig::instance().light_heartbeat_enabled() directly.

File: src/ray/raylet/node_manager.cc [hunk 4]
```diff
@@ -414,43 +404,56 @@ void NodeManager::Heartbeat() {
...
+        (*heartbeat_data->mutable_resources_available())[kResourcesTurningEmpty] = 0.0;
+      } else {
+        for (const auto &resource_pair :
+             local_resources.GetAvailableResources().GetResourceMap()) {
+          (*heartbeat_data->mutable_resources_available())[resource_pair.first] =
+              resource_pair.second;
+        }
+      }
+      last_heartbeat_resources_.SetAvailableResources(
+          ResourceSet(local_resources.GetAvailableResources()));
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:419]: How about abstract a function?

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 5]
```diff
@@ -459,9 +460,21 @@ void GcsNodeManager::StartNodeFailureDetector() {
...
     const ClientID &node_id, const rpc::HeartbeatTableData &heartbeat) {
-  auto resources_available = MapFromProtobuf(heartbeat.resources_available());
-  cluster_realtime_resources_[node_id] =
-      std::make_shared<ResourceSet>(resources_available);
+  if (light_heartbeat_enabled_) {
+    if (heartbeat.resources_available_size() > 0) {
+      if (heartbeat.resources_available().count(kResourcesTurningEmpty) == 1) {
+        cluster_realtime_resources_[node_id] = std::make_shared<ResourceSet>();
+      } else {
+        auto resources_available = MapFromProtobuf(heartbeat.resources_available());
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:468]: Based on your description, isn't `resources_available ` just an empty resourceSet?

File: src/ray/raylet/node_manager.cc [hunk 6]
```diff
@@ -414,43 +404,56 @@ void NodeManager::Heartbeat() {
...
       }
       last_heartbeat_resources_.SetLoadResources(
           ResourceSet(local_resources.GetLoadResources()));
     }
   } else {
     // If light heartbeat disabled, we send whole resources information every time.
     for (const auto &resource_pair :
-         local_resources.GetAvailableResources().GetResourceMap()) {
-      (*heartbeat_data->mutable_resources_available())[resource_pair.first] =
+         local_resources.GetTotalResources().GetResourceMap()) {
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:440]: Why did you switch these two?

File: src/ray/raylet/node_manager.cc [hunk 7]
```diff
@@ -414,43 +404,56 @@ void NodeManager::Heartbeat() {
...
+          (*heartbeat_data->mutable_resources_available())[resource_pair.first] =
+              resource_pair.second;
+        }
+      }
+      last_heartbeat_resources_.SetAvailableResources(
+          ResourceSet(local_resources.GetAvailableResources()));
+    }
+
     local_resources.SetLoadResources(local_queues_.GetTotalResourceLoad());
     if (!last_heartbeat_resources_.GetLoadResources().IsEqual(
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:423]: This code is too nested. Why don't we combine two predicates?

```
if (! last_heartbeat_resources_. GetLoadResources(). IsEqual. ... && local_resources.GetLoadResources().IsEmpty()) {
} else {
}
```

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:423]: Same for everywhere that has nested 

```
if (A) {
  if (B) {
  
  } else {}
}
```

=>

```
if (A && B) {

} else {}
```

File: src/ray/common/constants.h (hunk 1 above)
Comment (WangTaoTheTonic) [reply src/ray/common/constants.h:47]: ok

File: src/ray/gcs/gcs_server/gcs_node_manager.cc (hunk 2 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc:469]: ok

File: src/ray/gcs/gcs_server/gcs_node_manager.cc (hunk 3 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc]: ok

File: src/ray/raylet/node_manager.cc (hunk 4 above)
Comment (WangTaoTheTonic) [reply src/ray/raylet/node_manager.cc:419]: total / available / load resources are handled tiny differently. So a function might not be a proper way.

File: src/ray/gcs/gcs_server/gcs_node_manager.cc (hunk 5 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc]: Not always. If resource_available_size is not empty, it has two scenarios: 1. only contains "TURNING_EMPTY", which means available resources in this node become empty from non-empty. 2. contains number of resource key/value like { "cpu": 1, "gpu":2, "memory": 1024 }, this means available resources has changed in this node and it report newest available resources in heartbeat.

File: src/ray/raylet/node_manager.cc (hunk 6 above)
Comment (WangTaoTheTonic) [reply src/ray/raylet/node_manager.cc:440]: in elsewhere the codes normal handle total resources first, then available and load. I'd like keep them same order for readability.

File: src/ray/raylet/node_manager.cc (hunk 7 above)
Comment (WangTaoTheTonic) [reply src/ray/raylet/node_manager.cc:423]: No. It's not such a simple condition combination...
Conditions here are:
a) if A && B, do 1
b) if A && !B, do 2
c) if A, no matter what B is, do 3
d) if !A, no matter what B is, do nothing.
If we change the logic to
if (A && B) { do 1 } else { do 2}, we will miss c) and d)

File: src/ray/protobuf/gcs.proto [hunk 8]
```diff
@@ -281,16 +281,22 @@ message HeartbeatTableData {
   bytes client_id = 1;
   // Resource capacity currently available on this node manager.
   map<string, double> resources_available = 2;
+  // Indicates whether resources avaialbe is changed. Only used when
```
Comment (ffbin) [inline src/ray/protobuf/gcs.proto:284]:
```suggestion
  // Indicates whether avaialbe resources is changed. Only used when
```

**Review by ffbin — APPROVED** [review]
LGTM, just only small comment.

File: src/ray/raylet/node_manager.cc (hunk 7 above)
Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:420]: Ah, I see. That's right.

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Can you merge the latest master and fix lint?

**Comment by WangTaoTheTonic** [comment]
> Can you merge the latest master and fix lint?

Sorry that I thought the lint in gcs.proto is a false report. Now it's fixed and master codes are merged in.
It's good to go if you are ok with the new changes and ci is good. Thanks ~

**Comment by WangTaoTheTonic** [comment]
The tune test suite here: https://travis-ci.com/github/ray-project/ray/jobs/391327452 should not caused by this PR(I see it fails every where in other prs).

**Comment by rkooo567** [comment]
Windows build failed.

**Comment by WangTaoTheTonic** [comment]
> Windows build failed.

Fixed！

**Comment by rkooo567** [comment]
I like this commit lol 
```
@WangTaoTheTonic
fix embarrassed mistake
```

**Pull request merged**